hazelcast/src/test/java/com/hazelcast/test/PortAwareInstanceFactory.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.test;
+
+import static com.hazelcast.test.AbstractHazelcastClassRunner.getTestMethodName;
+import static com.hazelcast.test.HazelcastTestSupport.getAddress;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.config.JoinConfig;
+import com.hazelcast.config.TcpIpConfig;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+
+/**
+ * Per test-method factory for Hazelcast members. It sets first member host:port to TCP join configuration of other members. The
+ * instances are kept per test method which allows to terminate them in "After" methods (see {@link #terminateAll()}). The
+ * factory methods also sets custom group name which prevents accidental joins (e.g. dangling members).
+ */
```

Comment:
I would add that the usage of `{@link ParallelTest}` is allowed with this instance factory. This will make it easier to use the class, without analyzing the configuration details each time.

You could also use `{@link After}` in the JavaDoc, since it's in the test package, so JUnit is available.